File: ios-native/Dimo/Email/Gmail/GmailAPIClient.swift
import Foundation

struct GmailMessageReference: Decodable, Hashable, Sendable {
  var id: String
  var threadId: String?
}

struct GmailMessageListPage: Sendable {
  var messages: [GmailMessageReference]
  var nextPageToken: String?
}

struct GmailHistoryPage: Sendable {
  var addedMessageIds: [String]
  var latestHistoryId: String
  var nextPageToken: String?
}

struct GmailProfile: Decodable, Hashable, Sendable {
  var emailAddress: String
  var historyId: String
}

struct GmailMessageHeader: Decodable, Hashable, Sendable {
  var name: String
  var value: String
}

struct GmailMessageBody: Decodable, Hashable, Sendable {
  var attachmentId: String?
  var size: Int?
  var data: String?
}

struct GmailMessagePart: Decodable, Hashable, Sendable {
  var partId: String?
  var mimeType: String?
  var filename: String?
  var headers: [GmailMessageHeader]?
  var body: GmailMessageBody?
  var parts: [GmailMessagePart]?
}

struct GmailMessageResource: Decodable, Hashable, Sendable {
  var id: String
  var threadId: String
  var labelIds: [String]?
  var snippet: String?
  var historyId: String?
  var internalDate: String
  var payload: GmailMessagePart?
}

protocol GmailAPIClient: Sendable {
  func listMessages(
    subject: String,
    dimoUserId: String,
    since: Date,
    pageToken: String?
  ) async throws -> GmailMessageListPage

  func getMessages(
    subject: String,
    dimoUserId: String,
    ids: [String]
  ) async throws -> [GmailMessageResource]

  /// Fetches a MIME part body that Gmail returned as `attachmentId` instead of
  /// inline `data` (common for larger text/html bodies even without a file).
  func getAttachmentData(
    subject: String,
    dimoUserId: String,
    messageId: String,
    attachmentId: String
  ) async throws -> Data

  func listHistory(
    subject: String,
    dimoUserId: String,
    startHistoryId: String,
    pageToken: String?
  ) async throws -> GmailHistoryPage

  func profile(subject: String, dimoUserId: String) async throws -> GmailProfile
}

actor GmailRESTClient: GmailAPIClient {
  private static let apiBase = URL(string: "https://gmail.googleapis.com/gmail/v1")!
  private static let batchURL = URL(string: "https://gmail.googleapis.com/batch/gmail/v1")!
  private static let messageBatchSize = 20
  private static let batchPacingDelay: Duration = .milliseconds(1_500)
  private static let maximumRateLimitRetries = 6
  private let tokenProvider: any GmailAccessTokenProviding
  private let session: URLSession
  private let decoder = JSONDecoder()

  init(
    tokenProvider: any GmailAccessTokenProviding,
    session: URLSession = GmailURLSession.make()
  ) {
    self.tokenProvider = tokenProvider
    self.session = session
  }

  func listMessages(
    subject: String,
    dimoUserId: String,
    since: Date,
    pageToken: String?
  ) async throws -> GmailMessageListPage {
    let epochSeconds = Int(since.timeIntervalSince1970)
    var items = [
      URLQueryItem(name: "q", value: "after:\(epochSeconds) -in:spam -in:trash"),
      URLQueryItem(name: "includeSpamTrash", value: "false"),
      URLQueryItem(name: "maxResults", value: "100"),
    ]
    if let pageToken { items.append(URLQueryItem(name: "pageToken", value: pageToken)) }
    let url = try endpoint(path: "users/me/messages", queryItems: items)
    let data = try await authorizedData(
      request: URLRequest(url: url),
      subject: subject,
      dimoUserId: dimoUserId
    )
    let response = try decode(GmailMessageListResponse.self, from: data)
    return GmailMessageListPage(
      messages: response.messages ?? [],
      nextPageToken: response.nextPageToken
    )
  }

  func getMessages(
    subject: String,
    dimoUserId: String,
    ids: [String]
  ) async throws -> [GmailMessageResource] {
    var messages: [GmailMessageResource] = []
    let batches = ids.uniqued().chunked(maxCount: Self.messageBatchSize)
    for (index, batch) in batches.enumerated() {
      try Task.checkCancellation()
      var rateLimitAttempt = 0
      while true {
        do {
          messages += try await getMessageBatch(
            subject: subject,
            dimoUserId: dimoUserId,
            ids: batch
          )
          break
        } catch let GmailAPIError.rateLimited(retryAfter) {
          guard rateLimitAttempt < Self.maximumRateLimitRetries else { throw GmailAPIError.rateLimited(retryAfter: retryAfter) }
          try await waitForRateLimit(retryAfter: retryAfter, attempt: rateLimitAttempt)
          rateLimitAttempt += 1
        }
      }
      if index < batches.count - 1 {
        try await Task.sleep(for: Self.batchPacingDelay)
      }
    }
    return messages
  }

  func listHistory(
    subject: String,
    dimoUserId: String,
    startHistoryId: String,
    pageToken: String?
  ) async throws -> GmailHistoryPage {
    var items = [
      URLQueryItem(name: "startHistoryId", value: startHistoryId),
      URLQueryItem(name: "historyTypes", value: "messageAdded"),
      URLQueryItem(name: "maxResults", value: "100"),
    ]
    if let pageToken { items.append(URLQueryItem(name: "pageToken", value: pageToken)) }
    let url = try endpoint(path: "users/me/history", queryItems: items)
    do {
      let data = try await authorizedData(
        request: URLRequest(url: url),
        subject: subject,
        dimoUserId: dimoUserId
      )
      let response = try decode(GmailHistoryListResponse.self, from: data)
      let ids = (response.history ?? [])
        .flatMap { $0.messagesAdded ?? [] }
        .map(\.message.id)
        .uniqued()
      return GmailHistoryPage(
        addedMessageIds: ids,
        latestHistoryId: response.historyId,
        nextPageToken: response.nextPageToken
      )
    } catch let error as GmailAPIError {
      if case .httpStatus(404, _) = error { throw GmailAPIError.historyCursorExpired }
      throw error
    }
  }

  func profile(subject: String, dimoUserId: String) async throws -> GmailProfile {
    let url = try endpoint(path: "users/me/profile")
    let data = try await authorizedData(
      request: URLRequest(url: url),
      subject: subject,
      dimoUserId: dimoUserId
    )
    return try decode(GmailProfile.self, from: data)
  }

  func getAttachmentData(
    subject: String,
    dimoUserId: String,
    messageId: String,
    attachmentId: String
  ) async throws -> Data {
    let encodedMessage = messageId.addingPercentEncoding(withAllowedCharacters: .urlPathAllowed)
      ?? messageId
    let encodedAttachment = attachmentId.addingPercentEncoding(withAllowedCharacters: .urlPathAllowed)
      ?? attachmentId
    let url = try endpoint(
      path: "users/me/messages/\(encodedMessage)/attachments/\(encodedAttachment)"
    )
    let data = try await authorizedData(
      request: URLRequest(url: url),
      subject: subject,
      dimoUserId: dimoUserId
    )
    let response = try decode(GmailAttachmentResponse.self, from: data)
    guard let encoded = response.data,
          let decoded = GmailMessageParser.decodeBase64URLData(encoded)
    else {
      throw GmailAPIError.invalidResponse
    }
    return decoded
  }

  private func getMessageBatch(
    subject: String,
    dimoUserId: String,
    ids: [String],
    retriedAfterAuthentication: Bool = false
  ) async throws -> [GmailMessageResource] {
    guard !ids.isEmpty else { return [] }
    let boundary = "dimo_email_\(UUID().uuidString.replacingOccurrences(of: "-", with: ""))"
    var body = Data()
    for id in ids {
      let encodedId = id.addingPercentEncoding(withAllowedCharacters: .urlPathAllowed) ?? id
      // Build this byte-for-byte instead of using a multiline literal. Swift
      // strips the final source newline from a multiline string, which can
      // leave a bare CR immediately before the next MIME boundary.
      let part = "--\(boundary)\r\n"
        + "Content-Type: application/http\r\n"
        + "\r\n"
        + "GET /gmail/v1/users/me/messages/\(encodedId)?format=full HTTP/1.1\r\n"
        + "\r\n"
      body.append(Data(part.utf8))
    }
    body.append(Data("--\(boundary)--\r\n".utf8))

    let token = try await tokenProvider.accessToken(
      subject: subject,
      dimoUserId: dimoUserId,
      forceRefresh: retriedAfterAuthentication
    )
    var request = URLRequest(url: Self.batchURL)
    request.httpMethod = "POST"
    request.httpBody = body
    request.setValue("Bearer \(token.value)", forHTTPHeaderField: "Authorization")
    request.setValue("multipart/mixed; boundary=\(boundary)", forHTTPHeaderField: "Content-Type")
    let (data, response) = try await session.data(for: request)
    guard let http = response as? HTTPURLResponse else { throw GmailAPIError.invalidResponse }
    if http.statusCode == 401, !retriedAfterAuthentication {
      await tokenProvider.invalidate(subject: subject)
      return try await getMessageBatch(
        subject: subject,
        dimoUserId: dimoUserId,
        ids: ids,
        retriedAfterAuthentication: true
      )
    }
    try validate(http: http, data: data)
    guard let contentType = http.value(forHTTPHeaderField: "Content-Type") else {
      throw GmailAPIError.invalidBatchResponse
    }
    do {
      return try GmailBatchResponseParser.decodeMessages(
        data: data,
        contentType: contentType,
        decoder: decoder
      )
    } catch let GmailAPIError.httpStatus(status, _) where status == 401
      && !retriedAfterAuthentication {
      await tokenProvider.invalidate(subject: subject)
      return try await getMessageBatch(
        subject: subject,
        dimoUserId: dimoUserId,
        ids: ids,
        retriedAfterAuthentication: true
      )
    }
  }

  private func authorizedData(
    request: URLRequest,
    subject: String,
    dimoUserId: String,
    retriedAfterAuthentication: Bool = false,
    rateLimitAttempt: Int = 0
  ) async throws -> Data {
    let token = try await tokenProvider.accessToken(
      subject: subject,
      dimoUserId: dimoUserId,
      forceRefresh: retriedAfterAuthentication
    )
    var authorized = request
    authorized.setValue("Bearer \(token.value)", forHTTPHeaderField: "Authorization")
    let (data, response) = try await session.data(for: authorized)
    guard let http = response as? HTTPURLResponse else { throw GmailAPIError.invalidResponse }
    if http.statusCode == 401, !retriedAfterAuthentication {
      await tokenProvider.invalidate(subject: subject)
      return try await authorizedData(
        request: request,
        subject: subject,
        dimoUserId: dimoUserId,
        retriedAfterAuthentication: true,
        rateLimitAttempt: rateLimitAttempt
      )
    }
    do {
      try validate(http: http, data: data)
    } catch let GmailAPIError.rateLimited(retryAfter) {
      guard rateLimitAttempt < Self.maximumRateLimitRetries else {
        throw GmailAPIError.rateLimited(retryAfter: retryAfter)
      }
      try await waitForRateLimit(retryAfter: retryAfter, attempt: rateLimitAttempt)
      return try await authorizedData(
        request: request,
        subject: subject,
        dimoUserId: dimoUserId,
        retriedAfterAuthentication: retriedAfterAuthentication,
        rateLimitAttempt: rateLimitAttempt + 1
      )
    }
    return data
  }

  private func waitForRateLimit(retryAfter: TimeInterval?, attempt: Int) async throws {
    let exponentialDelay = min(32, pow(2, Double(attempt)))
    let requestedDelay = retryAfter.map { min(max($0, 1), 60) }
    let jitter = retryAfter == nil ? Double.random(in: 0..<0.75) : 0
    let delay = (requestedDelay ?? exponentialDelay) + jitter
    try await Task.sleep(for: .milliseconds(Int64(delay * 1_000)))
  }

  private func endpoint(path: String, queryItems: [URLQueryItem] = []) throws -> URL {
    let url = Self.apiBase.appending(path: path)
    guard var components = URLComponents(url: url, resolvingAgainstBaseURL: false) else {
      throw GmailAPIError.invalidRequest
    }
    if !queryItems.isEmpty { components.queryItems = queryItems }
    guard let result = components.url else { throw GmailAPIError.invalidRequest }
    return result
  }

  private func validate(http: HTTPURLResponse, data: Data) throws {
    guard !(200..<300).contains(http.statusCode) else { return }
    let retryAfter = http.value(forHTTPHeaderField: "Retry-After").flatMap(TimeInterval.init)
    if http.statusCode == 429 {
      throw GmailAPIError.rateLimited(retryAfter: retryAfter)
    }
    let envelope = try? decoder.decode(GmailErrorEnvelope.self, from: data)
    let reasons = envelope?.error.errors?.compactMap(\.reason) ?? []
    if http.statusCode == 403,
       reasons.contains(where: Self.isRateLimitReason) {
      throw GmailAPIError.rateLimited(retryAfter: retryAfter)
    }
    let message = envelope?.error.message
    throw GmailAPIError.httpStatus(http.statusCode, message)
  }

  private static func isRateLimitReason(_ reason: String) -> Bool {
    switch reason.lowercased() {
    case "ratelimitexceeded", "userratelimitexceeded", "quotaexceeded":
      return true
    default:
      return false
    }
  }

  private func decode<Value: Decodable>(_ type: Value.Type, from data: Data) throws -> Value {
    do { return try decoder.decode(type, from: data) }
    catch { throw GmailAPIError.invalidResponse }
  }
}

private struct GmailAttachmentResponse: Decodable {
  var size: Int?
  var data: String?
}

private struct GmailMessageListResponse: Decodable {
  var messages: [GmailMessageReference]?
  var nextPageToken: String?
}

private struct GmailHistoryListResponse: Decodable {
  struct History: Decodable {
    struct Added: Decodable { var message: GmailMessageReference }
    var messagesAdded: [Added]?
  }

  var history: [History]?
  var nextPageToken: String?
  var historyId: String
}

private struct GmailErrorEnvelope: Decodable {
  struct Body: Decodable {
    struct Detail: Decodable { var reason: String? }
    var message: String?
    var errors: [Detail]?
  }
  var error: Body
}

private enum GmailBatchResponseParser {
  static func decodeMessages(
    data: Data,
    contentType: String,
    decoder: JSONDecoder
  ) throws -> [GmailMessageResource] {
    guard let boundary = boundary(from: contentType),
          let response = String(data: data, encoding: .utf8) else {
      throw GmailAPIError.invalidBatchResponse
    }
    let delimiter = "--\(boundary)"
    var messages: [GmailMessageResource] = []
    for rawPart in response.components(separatedBy: delimiter) {
      let part = rawPart.trimmingCharacters(in: .whitespacesAndNewlines)
      guard !part.isEmpty, part != "--" else { continue }
      guard let statusRange = part.range(
        of: #"HTTP/(?:1\.1|2)\s+(\d{3})"#,
        options: .regularExpression
      ) else { throw GmailAPIError.invalidBatchResponse }
      let statusLine = String(part[statusRange])
      guard let status = statusLine.split(separator: " ").last.flatMap({ Int($0) }) else {
        throw GmailAPIError.invalidBatchResponse
      }
      if status == 404 { continue } // Message was deleted between list and get.
      if status == 429 { throw GmailAPIError.rateLimited(retryAfter: nil) }
      if status == 403 {
        let normalized = part.lowercased()
        if normalized.contains("ratelimitexceeded")
          || normalized.contains("userratelimitexceeded")
          || normalized.contains("quotaexceeded") {
          throw GmailAPIError.rateLimited(retryAfter: nil)
        }
      }
      guard (200..<300).contains(status) else {
        throw GmailAPIError.httpStatus(status, nil)
      }
      let afterStatus = part[statusRange.upperBound...]
      guard let headerEnd = afterStatus.range(of: "\r\n\r\n")
        ?? afterStatus.range(of: "\n\n") else {
        throw GmailAPIError.invalidBatchResponse
      }
      let json = afterStatus[headerEnd.upperBound...]
        .trimmingCharacters(in: .whitespacesAndNewlines)
      guard let jsonData = json.data(using: .utf8) else {
        throw GmailAPIError.invalidBatchResponse
      }
      do {
        messages.append(try decoder.decode(GmailMessageResource.self, from: jsonData))
      } catch {
        throw GmailAPIError.invalidBatchResponse
      }
    }
    return messages
  }

  private static func boundary(from contentType: String) -> String? {
    for component in contentType.split(separator: ";").dropFirst() {
      let pieces = component.split(separator: "=", maxSplits: 1)
      guard pieces.count == 2,
            pieces[0].trimmingCharacters(in: .whitespacesAndNewlines).lowercased() == "boundary"
      else { continue }
      return pieces[1].trimmingCharacters(in: CharacterSet(charactersIn: " \t\""))
    }
    return nil
  }
}

enum GmailAPIError: LocalizedError, Sendable {
  case invalidRequest
  case invalidResponse
  case invalidBatchResponse
  case historyCursorExpired
  case rateLimited(retryAfter: TimeInterval?)
  case httpStatus(Int, String?)

  var errorDescription: String? {
    switch self {
    case .invalidRequest: return "The Gmail API request was invalid."
    case .invalidResponse: return "Gmail returned an invalid response."
    case .invalidBatchResponse: return "Gmail returned an invalid batch response."
    case .historyCursorExpired: return "Gmail history expired; a new email scan is required."
    case .rateLimited: return "Gmail temporarily limited requests."
    case .httpStatus(let status, let message):
      return message.map { "Gmail returned HTTP \(status): \($0)" } ?? "Gmail returned HTTP \(status)."
    }
  }
}

private extension Array where Element: Hashable {
  func uniqued() -> [Element] {
    var seen = Set<Element>()
    return filter { seen.insert($0).inserted }
  }
}

private extension Array {
  func chunked(maxCount: Int) -> [[Element]] {
    guard maxCount > 0 else { return [] }
    return stride(from: 0, to: count, by: maxCount).map {
      Array(self[$0..<Swift.min($0 + maxCount, count)])
    }
  }
}

File: android-native/app/src/main/java/app/dimo/android/email/gmail/GmailAPIClient.kt
package app.dimo.android.email.gmail

import java.util.UUID
import kotlin.math.min
import kotlin.math.pow
import kotlin.random.Random
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.ensureActive
import kotlinx.coroutines.withContext
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import okhttp3.HttpUrl.Companion.toHttpUrl
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import kotlin.coroutines.coroutineContext

/**
 * Port of `ios-native/Dimo/Email/Gmail/GmailAPIClient.swift`.
 *
 * Only read paths exist. Message fetches go through Gmail's multipart batch
 * endpoint 20 at a time with pacing between batches, because per-message GETs
 * exhaust the per-user quota on any real inbox.
 */

@Serializable
data class GmailMessageReference(
  val id: String,
  val threadId: String? = null,
)

data class GmailMessageListPage(
  val messages: List<GmailMessageReference>,
  val nextPageToken: String?,
)

data class GmailHistoryPage(
  val addedMessageIds: List<String>,
  val latestHistoryId: String,
  val nextPageToken: String?,
)

@Serializable
data class GmailProfile(
  val emailAddress: String,
  val historyId: String,
)

@Serializable
data class GmailMessageHeader(
  val name: String,
  val value: String,
)

@Serializable
data class GmailMessageBody(
  val attachmentId: String? = null,
  val size: Int? = null,
  val data: String? = null,
)

@Serializable
data class GmailMessagePart(
  val partId: String? = null,
  val mimeType: String? = null,
  val filename: String? = null,
  val headers: List<GmailMessageHeader>? = null,
  val body: GmailMessageBody? = null,
  val parts: List<GmailMessagePart>? = null,
)

@Serializable
data class GmailMessageResource(
  val id: String,
  val threadId: String,
  val labelIds: List<String>? = null,
  val snippet: String? = null,
  val historyId: String? = null,
  val internalDate: String,
  val payload: GmailMessagePart? = null,
)

interface GmailAPIClient {
  suspend fun listMessages(
    subject: String,
    dimoUserId: String,
    since: Long,
    pageToken: String?,
  ): GmailMessageListPage

  suspend fun getMessages(
    subject: String,
    dimoUserId: String,
    ids: List<String>,
  ): List<GmailMessageResource>

  /**
   * Fetches a MIME part body Gmail returned as `attachmentId` instead of inline
   * `data` (common for larger text/html bodies even without a file).
   */
  suspend fun getAttachmentData(
    subject: String,
    dimoUserId: String,
    messageId: String,
    attachmentId: String,
  ): ByteArray

  suspend fun listHistory(
    subject: String,
    dimoUserId: String,
    startHistoryId: String,
    pageToken: String?,
  ): GmailHistoryPage

  suspend fun profile(subject: String, dimoUserId: String): GmailProfile
}

class GmailRESTClient(
  private val tokenProvider: GmailAccessTokenProviding,
  private val http: OkHttpClient = GmailHttp.client,
) : GmailAPIClient {
  private val json = Json { ignoreUnknownKeys = true }

  override suspend fun listMessages(
    subject: String,
    dimoUserId: String,
    since: Long,
    pageToken: String?,
  ): GmailMessageListPage {
    val url = endpoint("users/me/messages") {
      addQueryParameter("q", "after:${since / 1000} -in:spam -in:trash")
      addQueryParameter("includeSpamTrash", "false")
      addQueryParameter("maxResults", "100")
      if (pageToken != null) addQueryParameter("pageToken", pageToken)
    }
    val body = authorizedGet(url, subject, dimoUserId)
    val response = decode<GmailMessageListResponse>(body)
    return GmailMessageListPage(response.messages.orEmpty(), response.nextPageToken)
  }

  override suspend fun getMessages(
    subject: String,
    dimoUserId: String,
    ids: List<String>,
  ): List<GmailMessageResource> {
    val messages = mutableListOf<GmailMessageResource>()
    val batches = ids.distinct().chunked(MESSAGE_BATCH_SIZE)
    batches.forEachIndexed { index, batch ->
      coroutineContext.ensureActive()
      var rateLimitAttempt = 0
      while (true) {
        try {
          messages += getMessageBatch(subject, dimoUserId, batch)
          break
        } catch (error: GmailAPIException.RateLimited) {
          if (rateLimitAttempt >= MAX_RATE_LIMIT_RETRIES) throw error
          waitForRateLimit(error.retryAfterSeconds, rateLimitAttempt)
          rateLimitAttempt += 1
        }
      }
      if (index < batches.size - 1) delay(BATCH_PACING_MS)
    }
    return messages
  }

  override suspend fun listHistory(
    subject: String,
    dimoUserId: String,
    startHistoryId: String,
    pageToken: String?,
  ): GmailHistoryPage {
    val url = endpoint("users/me/history") {
      addQueryParameter("startHistoryId", startHistoryId)
      addQueryParameter("historyTypes", "messageAdded")
      addQueryParameter("maxResults", "100")
      if (pageToken != null) addQueryParameter("pageToken", pageToken)
    }
    val body = try {
      authorizedGet(url, subject, dimoUserId)
    } catch (error: GmailAPIException.HttpStatus) {
      // Gmail drops history older than about a week; the caller must re-scan.
      if (error.status == 404) throw GmailAPIException.HistoryCursorExpired else throw error
    }
    val response = decode<GmailHistoryListResponse>(body)
    val ids = response.history.orEmpty()
      .flatMap { it.messagesAdded.orEmpty() }
      .map { it.message.id }
      .distinct()
    return GmailHistoryPage(ids, response.historyId, response.nextPageToken)
  }

  override suspend fun profile(subject: String, dimoUserId: String): GmailProfile =
    decode(authorizedGet(endpoint("users/me/profile") {}, subject, dimoUserId))

  override suspend fun getAttachmentData(
    subject: String,
    dimoUserId: String,
    messageId: String,
    attachmentId: String,
  ): ByteArray {
    val url = API_BASE.toHttpUrl().newBuilder()
      .addPathSegments("users/me/messages")
      .addPathSegment(messageId)
      .addPathSegment("attachments")
      .addPathSegment(attachmentId)
      .build()
      .toString()
    val response = decode<GmailAttachmentResponse>(authorizedGet(url, subject, dimoUserId))
    return response.data?.let { GmailMessageParser.decodeBase64Url(it) }
      ?: throw GmailAPIException.InvalidResponse
  }

  // MARK: - Private

  private suspend fun getMessageBatch(
    subject: String,
    dimoUserId: String,
    ids: List<String>,
    retriedAfterAuthentication: Boolean = false,
  ): List<GmailMessageResource> {
    if (ids.isEmpty()) return emptyList()
    val boundary = "dimo_email_" + UUID.randomUUID().toString().replace("-", "")
    // Built byte-for-byte: a stray newline before a MIME boundary makes Gmail
    // reject the whole batch.
    val body = buildString {
      for (id in ids) {
        append("--").append(boundary).append("\r\n")
        append("Content-Type: application/http\r\n")
        append("\r\n")
        append("GET /gmail/v1/users/me/messages/").append(id).append("?format=full HTTP/1.1\r\n")
        append("\r\n")
      }
      append("--").append(boundary).append("--\r\n")
    }

    val token = tokenProvider.accessToken(subject, dimoUserId, retriedAfterAuthentication)
    val mediaType = "multipart/mixed; boundary=$boundary".toMediaType()
    val request = Request.Builder()
      .url(BATCH_URL)
      .post(body.toByteArray(Charsets.UTF_8).toRequestBody(mediaType))
      .header("Authorization", "Bearer ${token.value}")
      .build()

    val (status, contentType, payload) = withContext(Dispatchers.IO) {
      http.newCall(request).execute().use { response ->
        Triple(
          response.code,
          response.header("Content-Type"),
          response.body.string(),
        )
      }
    }
    if (status == 401 && !retriedAfterAuthentication) {
      tokenProvider.invalidate(subject)
      return getMessageBatch(subject, dimoUserId, ids, retriedAfterAuthentication = true)
    }
    validate(status, payload, retryAfterHeader = null)
    if (contentType == null) throw GmailAPIException.InvalidBatchResponse
    return try {
      GmailBatchResponseParser.decodeMessages(payload, contentType, json)
    } catch (error: GmailAPIException.HttpStatus) {
      if (error.status == 401 && !retriedAfterAuthentication) {
        tokenProvider.invalidate(subject)
        getMessageBatch(subject, dimoUserId, ids, retriedAfterAuthentication = true)
      } else {
        throw error
      }
    }
  }

  private suspend fun authorizedGet(
    url: String,
    subject: String,
    dimoUserId: String,
    retriedAfterAuthentication: Boolean = false,
    rateLimitAttempt: Int = 0,
  ): String {
    val token = tokenProvider.accessToken(subject, dimoUserId, retriedAfterAuthentication)
    val request = Request.Builder()
      .url(url)
      .header("Authorization", "Bearer ${token.value}")
      .build()
    val (status, retryAfter, body) = withContext(Dispatchers.IO) {
      http.newCall(request).execute().use { response ->
        Triple(
          response.code,
          response.header("Retry-After")?.toDoubleOrNull(),
          response.body.string(),
        )
      }
    }
    if (status == 401 && !retriedAfterAuthentication) {
      tokenProvider.invalidate(subject)
      return authorizedGet(url, subject, dimoUserId, true, rateLimitAttempt)
    }
    try {
      validate(status, body, retryAfter)
    } catch (error: GmailAPIException.RateLimited) {
      if (rateLimitAttempt >= MAX_RATE_LIMIT_RETRIES) throw error
      waitForRateLimit(error.retryAfterSeconds, rateLimitAttempt)
      return authorizedGet(
        url,
        subject,
        dimoUserId,
        retriedAfterAuthentication,
        rateLimitAttempt + 1,
      )
    }
    return body
  }

  private suspend fun waitForRateLimit(retryAfterSeconds: Double?, attempt: Int) {
    val exponential = min(32.0, 2.0.pow(attempt))
    val requested = retryAfterSeconds?.coerceIn(1.0, 60.0)
    // Jitter only when the server did not name a delay, so a fleet of clients
    // does not retry in lockstep.
    val jitter = if (retryAfterSeconds == null) Random.nextDouble(0.0, 0.75) else 0.0
    delay(((requested ?: exponential) + jitter).times(1000).toLong())
  }

  private fun endpoint(path: String, build: okhttp3.HttpUrl.Builder.() -> Unit): String =
    API_BASE.toHttpUrl().newBuilder().addPathSegments(path).apply(build).build().toString()

  private fun validate(status: Int, body: String, retryAfterHeader: Double?) {
    if (status in 200..299) return
    if (status == 429) throw GmailAPIException.RateLimited(retryAfterHeader)
    val envelope = runCatching { json.decodeFromString<GmailErrorEnvelope>(body) }.getOrNull()
    val reasons = envelope?.error?.errors.orEmpty().mapNotNull { it.reason }
    // Gmail reports quota exhaustion as 403 with a reason, not as 429.
    if (status == 403 && reasons.any(::isRateLimitReason)) {
      throw GmailAPIException.RateLimited(retryAfterHeader)
    }
    throw GmailAPIException.HttpStatus(status, envelope?.error?.message)
  }

  private fun isRateLimitReason(reason: String): Boolean = when (reason.lowercase()) {
    "ratelimitexceeded", "userratelimitexceeded", "quotaexceeded" -> true
    else -> false
  }

  private inline fun <reified T> decode(body: String): T =
    runCatching { json.decodeFromString<T>(body) }.getOrNull()
      ?: throw GmailAPIException.InvalidResponse

  private companion object {
    const val API_BASE = "https://gmail.googleapis.com/gmail/v1"
    const val BATCH_URL = "https://gmail.googleapis.com/batch/gmail/v1"
    const val MESSAGE_BATCH_SIZE = 20
    const val BATCH_PACING_MS = 1_500L
    const val MAX_RATE_LIMIT_RETRIES = 6
  }
}

@Serializable
private data class GmailAttachmentResponse(
  val size: Int? = null,
  val data: String? = null,
)

@Serializable
private data class GmailMessageListResponse(
  val messages: List<GmailMessageReference>? = null,
  val nextPageToken: String? = null,
)

@Serializable
private data class GmailHistoryListResponse(
  val history: List<History>? = null,
  val nextPageToken: String? = null,
  val historyId: String,
) {
  @Serializable
  data class History(val messagesAdded: List<Added>? = null)

  @Serializable
  data class Added(val message: GmailMessageReference)
}

@Serializable
private data class GmailErrorEnvelope(val error: Body) {
  @Serializable
  data class Body(
    val message: String? = null,
    val errors: List<Detail>? = null,
  )

  @Serializable
  data class Detail(val reason: String? = null)
}

/** Splits Gmail's `multipart/mixed` batch response into the individual JSON bodies. */
internal object GmailBatchResponseParser {
  fun decodeMessages(
    response: String,
    contentType: String,
    json: Json,
  ): List<GmailMessageResource> {
    val boundary = boundary(contentType) ?: throw GmailAPIException.InvalidBatchResponse
    val messages = mutableListOf<GmailMessageResource>()
    for (rawPart in response.split("--$boundary")) {
      val part = rawPart.trim()
      if (part.isEmpty() || part == "--") continue
      val statusMatch = STATUS_LINE.find(part) ?: throw GmailAPIException.InvalidBatchResponse
      val status = statusMatch.groupValues[1].toIntOrNull()
        ?: throw GmailAPIException.InvalidBatchResponse
      // Deleted between list and get; not an error for this sync.
      if (status == 404) continue
      if (status == 429) throw GmailAPIException.RateLimited(null)
      if (status == 403) {
        val normalized = part.lowercase()
        if (normalized.contains("ratelimitexceeded") ||
          normalized.contains("userratelimitexceeded") ||
          normalized.contains("quotaexceeded")
        ) {
          throw GmailAPIException.RateLimited(null)
        }
      }
      if (status !in 200..299) throw GmailAPIException.HttpStatus(status, null)

      val afterStatus = part.substring(statusMatch.range.last + 1)
      val headerEnd = afterStatus.indexOf("\r\n\r\n").takeIf { it >= 0 }?.let { it + 4 }
        ?: afterStatus.indexOf("\n\n").takeIf { it >= 0 }?.let { it + 2 }
        ?: throw GmailAPIException.InvalidBatchResponse
      val body = afterStatus.substring(headerEnd).trim()
      messages += runCatching { json.decodeFromString<GmailMessageResource>(body) }.getOrNull()
        ?: throw GmailAPIException.InvalidBatchResponse
    }
    return messages
  }

  private fun boundary(contentType: String): String? {
    for (component in contentType.split(";").drop(1)) {
      val pieces = component.split("=", limit = 2)
      if (pieces.size != 2) continue
      if (pieces[0].trim().lowercase() != "boundary") continue
      return pieces[1].trim().trim('"', ' ', '\t')
    }
    return null
  }

  private val STATUS_LINE = Regex("HTTP/(?:1\\.1|2)\\s+(\\d{3})")
}

sealed class GmailAPIException(message: String) : Exception(message) {
  object InvalidRequest : GmailAPIException("The Gmail API request was invalid.")
  object InvalidResponse : GmailAPIException("Gmail returned an invalid response.")
  object InvalidBatchResponse : GmailAPIException("Gmail returned an invalid batch response.")
  object HistoryCursorExpired :
    GmailAPIException("Gmail history expired; a new email scan is required.")

  class RateLimited(val retryAfterSeconds: Double?) :
    GmailAPIException("Gmail temporarily limited requests.")

  class HttpStatus(val status: Int, val detail: String?) : GmailAPIException(
    if (detail != null) "Gmail returned HTTP $status: $detail" else "Gmail returned HTTP $status.",
  )
}
